biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts

```diff
@@ -5,62 +5,176 @@ import {
-      town.getTownSize(),
-      opts.numberOfLargeBatteryCells
-    );
-    const smallBatteries = smallBatteryPositions.map(
-      (pos) => new BioBattery(pos.x, pos.y)
-    );
-    const largeBatteries = largeBatteryPositions.map(
-      (pos) => new BioBattery(pos.x, pos.y)
-    );
-    this.state = new BiogridState([
-      ...smallBatteries,
-      ...largeBatteries,
-      ...town.getEnergyUsers(),
-    ]);
-    this.state.convertStateGraphToMST();
+
+    // Batteries
+    // TODO implement this outside when calling the grid
+    const smallBatteryPositions = this.createGridItemPositions(town.getTownSize(), opts.numberOfSmallBatteryCells);
+    const largeBatteryPositions = this.createGridItemPositions(town.getTownSize(), opts.numberOfLargeBatteryCells);
+    // TODO constants
+    this.smallBatteries = this.createBatteries(smallBatteryPositions, SMALL_BATTERY.DEFAULT_START_ENERGY, SMALL_BATTERY.MAX_CAPACITY, GRID_ITEM_NAMES.SMALL_BATTERY);
+    this.largeBatteries = this.createBatteries(largeBatteryPositions, LARGE_BATTERY.DEFAULT_START_ENERGY, LARGE_BATTERY.MAX_CAPACITY, GRID_ITEM_NAMES.LARGE_BATTERY);
+
+    // Towns
+    this.town = town;
+
+    // Enery Source
+    // TODO implement the solar panels
+    const solarPanelPositions = this.createGridItemPositions(town.getTownSize(), opts.numberOfSolarPanels);
+    this.solarPanels = this.createSolarPanels(solarPanelPositions);
+
+    this.state = new BiogridState(this.createGridItems());
+
   }
 
-  getSystemState(): BiogridState {
+  private createGridItems(): GridItem[] {
+    return [
+      ...this.smallBatteries,
+      ...this.largeBatteries,
+      ...this.town.getEnergyUsers(),
+      ...this.solarPanels,
+    ]
+  }
+
+  getSystemState() {
     return this.state;
   }
 
+  getJsonGraphDetails() {
+    this.state.getJsonGraph();
+  }
+
+  private createBatteries(positions: ItemPosition[], initEnergy: Energy, maxCapacity: Energy, name: string): Battery[] {
+    return positions.map((position, index) => new BioBattery(position.x, position.y, `${name}-${index}`, initEnergy, maxCapacity));
+  }
+
+  private createSolarPanels(positions: ItemPosition[]): EnergySource[] {
+    return positions.map((position, index) => new BioEnergySource(position.x, position.y, `${GRID_ITEM_NAMES.SOLAR_PANEL}-${index}`, SOLAR_PANEL.DEFAULT_INITIAL_ENERGY, SOLAR_PANEL.MIN_CAPACITY))
+  }
+
   takeAction(action: GridAction) {
-    // TODO implement actions taken later
-    return;
+    // RETURN a new BiogridState
+    const allSupplyingPaths = action.getSupplyingPaths()
+
+    const clonedGraph = this.state.cloneStateGraph();
+
+    for (const supplyPath in allSupplyingPaths) {
+      const oldGridItem = this.state.getGridItem(supplyPath);
+      const supplyingGridItem = this.state.getGridItem(allSupplyingPaths[supplyPath]);
+      const typeOldGridItem = this.getGridItemType(oldGridItem);
+      if (typeOldGridItem === GRID_ITEM_NAMES.ENERGY_USER) {
```

Comment:
could this be broken up using helper functions. I.e. if it is an energy user, a function like `takeEnergyUserAction`

Reply:
> could this be broken up using helper functions. I.e. if it is an energy user, a function like `takeEnergyUserAction`

will look at how it can be done like this. Thanks for highlighting this out